pkg/bloomcompactor/retention.go

```diff
@@ -0,0 +1,286 @@
+				tenantRetention := findLongestRetention(globalRetention, streamRetention)
+				expirationDay := storageconfig.NewDayTime(today.Add(-tenantRetention))
+				return day.Before(expirationDay)
+			},
+		)
+		if err != nil {
+			r.metrics.retentionTime.WithLabelValues(statusFailure).Observe(time.Since(start.Time()).Seconds())
+			r.metrics.retentionDaysPerIteration.WithLabelValues(statusFailure).Observe(float64(daysProcessed))
+			r.metrics.retentionTenantsPerIteration.WithLabelValues(statusFailure).Observe(float64(len(tenantsRetentionApplied)))
+			return errors.Wrap(err, "getting users for period")
+		}
+
+		if len(tenants) == 0 {
+			// No tenants for this day means we can break here since previous
+			// retention iterations have already deleted all tenants
+			break
+		}
+
+		for tenant, objects := range tenants {
+			if len(objects) == 0 {
+				continue
+			}
+
+			tenantLogger := log.With(dayLogger, "tenant", tenant)
+			level.Info(tenantLogger).Log("msg", "applying retention to tenant", "keys", len(objects))
+
+			// Note: we cannot delete the tenant directory directly because it is not an
+			// actual key in the object store. Instead, we need to delete all keys one by one.
+			for _, object := range objects {
+				if err := objectClient.DeleteObject(ctx, object.Key); err != nil {
+					r.metrics.retentionTime.WithLabelValues(statusFailure).Observe(time.Since(start.Time()).Seconds())
+					r.metrics.retentionDaysPerIteration.WithLabelValues(statusFailure).Observe(float64(daysProcessed))
+					r.metrics.retentionTenantsPerIteration.WithLabelValues(statusFailure).Observe(float64(len(tenantsRetentionApplied)))
+					return errors.Wrapf(err, "deleting key %s", object.Key)
+				}
+			}
+
+			tenantsRetentionApplied[tenant] = struct{}{}
+		}
+
+		daysProcessed++
+	}
+
+	r.lastDayRun = today
+	r.metrics.retentionTime.WithLabelValues(statusSuccess).Observe(time.Since(start.Time()).Seconds())
+	r.metrics.retentionDaysPerIteration.WithLabelValues(statusSuccess).Observe(float64(daysProcessed))
+	r.metrics.retentionTenantsPerIteration.WithLabelValues(statusSuccess).Observe(float64(len(tenantsRetentionApplied)))
+	level.Info(r.logger).Log("msg", "finished applying retention", "daysProcessed", daysProcessed, "tenants", len(tenantsRetentionApplied))
+
+	return nil
+}
+
+func findLongestRetention(globalRetention time.Duration, streamRetention []validation.StreamRetention) time.Duration {
+	if len(streamRetention) == 0 {
+		return globalRetention
+	}
+
+	maxStreamRetention := slices.MaxFunc(streamRetention, func(a, b validation.StreamRetention) int {
+		return int(a.Period - b.Period)
+	})
+
+	if time.Duration(maxStreamRetention.Period) > globalRetention {
+		return time.Duration(maxStreamRetention.Period)
+	}
+	return globalRetention
+}
+
+// findSmallestRetention returns the smallest retention period across all tenants.
+// It also returns a boolean indicating if there is any retention period set at all
```

Comment:
This is somewhat confusing. It returns the smallest retention, but skips zero (disabled). It also does not return a bool like it suggests. Maybe `smallestEnabledRetention`?

Reply:
Changed, thanks.